treeherder/auth/backends.py

```diff
@@ -0,0 +1,148 @@
+import logging
+import re
+from hashlib import sha1
+
+from django.contrib.auth.models import User
+from django.core.exceptions import ObjectDoesNotExist
+from django.db.models import Q
+from taskcluster.sync import Auth
+from taskcluster.utils import scope_match
+
+logger = logging.getLogger(__name__)
+
+
+class TaskclusterAuthBackend(object):
+    """
+        result of tc_auth.authenticateHawk has the form:
+
+        {'status': 'auth-success',
+         'scopes': ['assume:mozilla-group:ateam',
+                    'assume:mozilla-group:vpn_treeherder',
+                    'assume:mozilla-user:biped@mozilla.com',
+                    'assume:mozillians-user:biped',
+                    ...
+                    'assume:project-admin:ateam',
+                    'assume:project-admin:treeherder',
+                    'assume:project:ateam:*',
+                    'assume:project:treeherder:*',
+                    'assume:worker-id:*',
+                    'secrets:set:project/treeherder/*'],
+         'scheme': 'hawk',
+         'clientId': 'mozilla-ldap/biped@mozilla.com',
+         'expires': '2016-10-31T17:40:45.692Z'}
+    """
+
+    def _get_scope_value(self, result, scope_prefix):
+        for scope in result["scopes"]:
+            if scope.startswith(scope_prefix):
+                return scope[len(scope_prefix):]
+        return None
+
+    def _get_email(self, result):
+        """
+        Get the user's email from the mozilla-user scope
+
+        """
+        email = self._get_scope_value(result, "assume:mozilla-user:")
+
+        if not email or email == "*":
+            # try finding the email in the clientId.  It can then be used to
```

Comment:
As I understand it, ``assume:mozilla-user:...`` will either be not present, have an email, or have * if the user is an admin.  So, yeah, that's the case this is trying to handle.  Dustin had talked about failing authentication in that case, thereby forcing the TC team to fix it on their end.  We can do that, for sure.

I'm not sure I follow your comment about "If someone can spoof the assume:mozilla-user:... scope".  If we can't trust the ``clientId`` and ``scopes`` that come back from the response, then we can't trust any of this anyway, right?   

If a user tries to use Persona, then they wouldn't have that ``mozilla-user`` scope.  But that's going away anyway, so not really an issue.

Though, you DO bring up a point that Dustin had brought up a while back and I'd forgotten about:  Trying to find a user with the truncated username (due to 30 char limit) could give the wrong user if the email address in their clientId was super long, making the trail end of it be the same for two different people. 

That being said, you're right.  We should just fail if we don't have a valid email from that scope.